src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -53,29 +53,75 @@ void ChunkPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) co
-    stored_table->set_pruned_chunk_ids(intersection);
-  } else {
-    stored_table->set_pruned_chunk_ids(std::vector<ChunkID>(pruned_chunk_ids.begin(), pruned_chunk_ids.end()));
+  // Gather PredicateNodes on top of a StoredTableNode. Ignore non-filtering, ValidateNodes and Joins.
+  // Starting from the current node, drilling down might result in multple cains of predicates ending
+  // at different or the same StoredTableNodes.
+  // predicates_for_table_nodes keeps track of the predicate chains for each stored table
+  std::unordered_map<std::shared_ptr<StoredTableNode>, std::vector<PredicateChain>> predicates_for_table_nodes{};
+  _recurse_on_inputs(node, predicates_for_table_nodes, PredicateChain{});
+
+  for (auto& [stored_table, predicate_chains] : predicates_for_table_nodes) {
+    /**
+     * Chains of predicates followed by a stored table node were found.
+     */
+    auto table = Hyrise::get().storage_manager.get_table(stored_table->table_name);
+
+    for (auto& predicate_chain : predicate_chains) {
+      std::set<ChunkID> pruned_chunk_ids;
+      for (auto& predicate : predicate_chain) {
+        // Predicate chains might contain predicates that are not applicable to the stored table
+        // e.g. operate on coulumns that originate from a different joined table
+        bool predicate_matches_table{true};
+        auto predicate_expression = predicate->predicate();
+        visit_expression(predicate_expression, [&](auto& expression_ptr) {
+          if (expression_ptr->type != ExpressionType::LQPColumn) return ExpressionVisitation::VisitArguments;
+
+          const auto lqp_column_expression_ptr = std::dynamic_pointer_cast<LQPColumnExpression>(expression_ptr);
+          Assert(lqp_column_expression_ptr,
+                 "Asked to adapt expression in LQP, but encountered non-LQP ColumnExpression");
+
+          if (lqp_column_expression_ptr->original_node.lock() != stored_table) {
+            predicate_matches_table = false;
+          }
+
+          return ExpressionVisitation::DoNotVisitArguments;
+        });
+
+        if (predicate_matches_table) {
+          auto new_exclusions = _compute_exclude_list(*table, *predicate->predicate(), stored_table);
+          pruned_chunk_ids.insert(new_exclusions.begin(), new_exclusions.end());
+        }
+      }
+
+      // wanted side effect of using sets: pruned_chunk_ids vector is sorted
+      const auto& already_pruned_chunk_ids = stored_table->pruned_chunk_ids();
+      if (!already_pruned_chunk_ids.empty()) {
+        std::vector<ChunkID> intersection;
+        std::set_intersection(already_pruned_chunk_ids.begin(), already_pruned_chunk_ids.end(),
+                              pruned_chunk_ids.begin(), pruned_chunk_ids.end(), std::back_inserter(intersection));
+        stored_table->set_pruned_chunk_ids(intersection);
+      } else {
+        stored_table->set_pruned_chunk_ids(std::vector<ChunkID>(pruned_chunk_ids.begin(), pruned_chunk_ids.end()));
+      }
+    }
```

Comment:
I'll need to look into this rule more deeply in the second pass. Feels like it could be possible to simplify it a bit.